File: emm/server_service/src/schemas/probe_settings.py
"""Pydantic-схемы настроек проб статуса (`/settings/probes`).

Платформенный singleton под `account_admin`. Описывает частоту и вкл/выкл двух
групп проб, которые снимает server_worker (reachability = ping+ssh,
power = ipmi/domstate).

Нижние границы интервалов проверяются на уровне полей (`ge`). Правило
«power >= reachability» здесь не выражается (PUT частичный — второе поле может
не приходить), его проверяет сервис на слитых эффективных значениях.
"""

from pydantic import BaseModel, Field

from src.models.probe_settings import (
    MIN_POWER_INTERVAL_SECONDS,
    MIN_REACHABILITY_INTERVAL_SECONDS,
)


class ProbeSettingsResponse(BaseModel):
    """Текущие настройки проб статуса."""

    reachability_probe_interval_seconds: int = Field(
        description="Интервал пробы доступности (ping+ssh), секунды.",
    )
    power_probe_interval_seconds: int = Field(
        description="Интервал пробы питания (ipmi/domstate), секунды.",
    )
    reachability_probe_enabled: bool = Field(
        description="Включена ли проба доступности.",
    )
    power_probe_enabled: bool = Field(
        description="Включена ли проба питания.",
    )


class ProbeSettingsUpdate(BaseModel):
    """Тело PUT — частичное обновление настроек проб.

    Любое поле можно опустить — тогда текущее значение сохраняется. Нижние
    границы интервалов enforce'ятся здесь; порядок (power >= reachability) —
    в сервисе на эффективных значениях после слияния с текущими.
    """

    reachability_probe_interval_seconds: int | None = Field(
        default=None,
        ge=MIN_REACHABILITY_INTERVAL_SECONDS,
        description=(
            "Интервал пробы доступности (ping+ssh), секунды. "
            f"Минимум {MIN_REACHABILITY_INTERVAL_SECONDS}."
        ),
    )
    power_probe_interval_seconds: int | None = Field(
        default=None,
        ge=MIN_POWER_INTERVAL_SECONDS,
        description=(
            "Интервал пробы питания (ipmi/domstate), секунды. "
            f"Минимум {MIN_POWER_INTERVAL_SECONDS}."
        ),
    )
    reachability_probe_enabled: bool | None = Field(
        default=None,
        description="Включить/выключить пробу доступности.",
    )
    power_probe_enabled: bool | None = Field(
        default=None,
        description="Включить/выключить пробу питания.",
    )

File: emm/web_ui/src/api/server/probeSettings.ts
/**
 * Настройки проб статуса `server_service`.
 *
 * Пробы статуса (reachability = ping+ssh, power = ipmi/domstate) снимает
 * server_worker. Их частота и вкл/выкл хранятся в БД (а не в env воркера),
 * чтобы работать одинаково в docker и k8s. Воркер читает настройки через
 * internal-эндпоинт; редактирует их account_admin.
 *
 * Источник истины — `server_service` endpoint'ы:
 *   GET  /api/server/v1/settings/probes
 *   PUT  /api/server/v1/settings/probes
 * Гейтится account_admin; остальным backend ответит 403.
 */

import { apiGet, apiPut } from "@/api/client";

/** Нижние границы интервалов (совпадают с backend'ом). */
export const MIN_REACHABILITY_INTERVAL_SECONDS = 15;
export const MIN_POWER_INTERVAL_SECONDS = 60;

/** Текущие настройки проб (ответ GET). */
export interface ProbeSettings {
  /** Интервал пробы доступности (ping+ssh), секунды. */
  reachability_probe_interval_seconds: number;
  /** Интервал пробы питания (ipmi/domstate), секунды. */
  power_probe_interval_seconds: number;
  /** Включена ли проба доступности. */
  reachability_probe_enabled: boolean;
  /** Включена ли проба питания. */
  power_probe_enabled: boolean;
}

/** Тело PUT — частичное обновление настроек проб. */
export type ProbeSettingsUpdate = Partial<ProbeSettings>;

const BASE = "/server/v1";

/** Прочитать текущие настройки проб. 403 — нет account_admin. */
export function getProbeSettings(): Promise<ProbeSettings> {
  return apiGet<ProbeSettings>(`${BASE}/settings/probes`);
}

/** Обновить настройки проб (частичная замена). */
export function putProbeSettings(
  body: ProbeSettingsUpdate,
): Promise<ProbeSettings> {
  return apiPut<ProbeSettings>(`${BASE}/settings/probes`, body);
}
